cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,225 @@
+              ValueType* values,
+              IndexType num_rows,
+              IndexType num_cols,
+              IndexType nnz,
+              raft::device_vector_view<IndexType, int64_t> idFeatCount,
+              int& fullFeatCount,
+              raft::device_vector_view<IndexType, int64_t> rowFeatCnts)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  int uniq_cnt  = raft::sparse::neighbors::get_n_components(rows, nnz, stream);
+  auto row_keys = raft::make_device_vector<IndexType>(handle, uniq_cnt);
+  auto row_cnts = raft::make_device_vector<ValueType>(handle, uniq_cnt);
+  get_uniques_counts<IndexType, ValueType, int64_t>(
+    handle, rows, columns, values, nnz, row_keys.view(), row_cnts.view());
+
+  auto dummy_vec = raft::make_device_vector<IndexType>(handle, uniq_cnt);
+  raft::linalg::map(handle,
+                    dummy_vec.view(),
+                    mapper<IndexType, ValueType>(rowFeatCnts.data_handle()),
+                    raft::make_const_mdspan(row_keys.view()),
+                    raft::make_const_mdspan(row_cnts.view()));
+
+  fit_tfidf(handle, columns, values, num_cols, nnz, idFeatCount, fullFeatCount);
+}
+
+/**
+ * @brief This function calculate the BM25 value for the COO records of the matrix.
+ * @param handle: raft resource handle
+ * @param rows: Input COO rows, of size nnz.
+ * @param columns: Input COO columns, of size nnz.
+ * @param values: Input COO values array, of size nnz.
+ * @param num_rows: total number of rows in the matrix.
+ * @param idFeatCount: array holding the occurrences per feature for matrix, size of num_cols.
+ * @param fullFeatCount: value corresponding to total number of features in matrix.
+ * @param rowFeatCnts array holding the feature occurrences per row for matrix, size of num_rows.
+ * @param k_param bm25 okapi optimization parameter k1.
+ * @param b_param bm25 okapi optimization parameter b.
+ * @param results Output array with values of the bm25 transform for each coo entry.
+ */
+template <typename ValueType = float, typename IndexType = int>
+void transform_bm25(raft::resources const& handle,
+                    IndexType* rows,
+                    IndexType* columns,
+                    ValueType* values,
+                    IndexType num_rows,
+                    raft::device_vector_view<IndexType, int64_t> featIdCount,
+                    IndexType fullIdLen,
+                    raft::device_vector_view<IndexType, int64_t> rowFeatCnts,
+                    float k_param,
+                    float b_param,
+                    raft::device_vector_view<ValueType, int64_t> results)
+{
+  float avgIdLen = (ValueType)fullIdLen / num_rows;
+  raft::linalg::map_offset(handle, results, [=] __device__(IndexType idx) {
+    ValueType tf     = (ValueType)raft::log<ValueType>(values[idx]);
+    ValueType idf_in = (ValueType)num_rows / featIdCount.data_handle()[columns[idx]];
+    ValueType idf    = (ValueType)raft::log<ValueType>(idf_in + 1);
```

Comment:
nitpick: We should stay away from C-style casts (like `(ValueType)` here) - and use the corresponding c++ casts (`static_cast<ValueType>(...)` etc) instead.

However - I don't think we need a cast in a couple of these cases, and it seems like we might be casting the same type to itself.  Maybe something like:

```suggestion
    ValueType tf     = raft::log<ValueType>(values[idx]);
    ValueType idf_in = static_cast<ValueType>(num_rows) / featIdCount.data_handle()[columns[idx]];
    ValueType idf    = raft::log<ValueType>(idf_in + 1);
```